packages/ab-testing/src/index.ts

```diff
@@ -1,9 +1,74 @@
-// add another package with `yo lerna-typescript:package @myscope/my-new-package` instead of `lerna create`
+import crc32 from 'fast-crc32c';
 
-export function greet(name: string): string {
-    return `ab-testing says: hello to ${name}`;
+type HashObject = (
+    object: { [key: string]: string | number | undefined },
+    salt: string
+) => { [s: string]: string };
+
+type ForceInclude = {
+    [hashedKey: string]: string[];
+};
+
+type Cohort = {
+    name: string;
+    allocation?: [number, number][];
+    force_include?: ForceInclude;
+};
+
+type Experiment = {
+    name: string;
+    cohorts: Cohort[];
+};
+
+export type ABTestingConfig = {
+    salt: string;
+    version: '1.0';
+    experiments: Experiment[];
+};
+
+export type UserProfile = {
+    user_id: number;
+    user_type?: string;
+    email?: string;
+    email_domain?: string;
+};
+
+function getModuloValue(experiment: string, userId: number): number {
+    return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
 }
 
-export function add(a: number, b: number): number {
-    return a + b;
+export class Experiments {
+    config: ABTestingConfig;
+    profile: UserProfile;
+    hashedProfile: ReturnType<HashObject>;
+
+    constructor(config: ABTestingConfig, profile: UserProfile, hashProfile: HashObject) {
+        this.config = config;
+        this.profile = profile;
+        this.hashedProfile = hashProfile(profile, config.salt);
+    }
+
+    getCohort = (experimentName: string): string => {
+        const experimentConfig = this.config.experiments.find(e => e.name === experimentName);
+        const userSegmentNum = getModuloValue(experimentName, this.profile.user_id);
+        let allocatedCohort = 'control';
+        for (const cohort of experimentConfig?.cohorts || []) {
+            for (const allocation of cohort.allocation || []) {
+                if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
+                    allocatedCohort = cohort.name;
+                }
+            }
+            if (cohort.force_include) {

```

Comment:
Let's put this case first, we don't need to loop allocation if we're going to force include a user. So this step should come first.

After that, during the allocation check let's return on the first match (this make sure we don't get many matches and select the later one.) That type of conflict is handled in our management tool, but might not be by other users of our library; so let's make sure we're as consistent as possible.

Reply:
neat, fixed